clients/drcachesim/tracer/tracer.cpp

```diff
@@ -211,16 +219,185 @@ static uint     tls_offs;
+        tmp = (buf_entry_t *)drvector_get_entry(&circular_buf_queue, queue_get);
+        *entry = *tmp;
+        init_buf_entry(tmp);
+        queue_get = (queue_get + 1) & queue_size_mask;
+        get = true;
+    }
+    drvector_unlock(&circular_buf_queue);
+    return get;
+}
+
+static void
+trace_traverse(byte *buf_base, ssize_t size,
+               bool vaddr2paddr, bool do_write, thread_id_t tid)
+{
+    byte *mem_ref, *buf_ptr;
+    byte *pipe_start, *pipe_end;
+    pipe_start = buf_base;
+    pipe_end = pipe_start;
+    buf_ptr = buf_base + size;
+    for (mem_ref  = buf_base + buf_hdr_slots_size;
+         mem_ref  < buf_ptr;
+         mem_ref += instru->sizeof_entry()) {
+        if (vaddr2paddr) {
+            trace_type_t type = instru->get_entry_type(mem_ref);
+            if (type != TRACE_TYPE_THREAD &&
+                type != TRACE_TYPE_THREAD_EXIT &&
+                type != TRACE_TYPE_PID) {
+                addr_t virt = instru->get_entry_addr(mem_ref);
+                addr_t phys = physaddr.virtual2physical(virt);
+                DR_ASSERT(type != TRACE_TYPE_INSTR_BUNDLE);
+                if (phys != 0)
+                    instru->set_entry_addr(mem_ref, phys);
+                else {
+                    // XXX i#1735: use virtual address and continue?
+                    // There are cases the xl8 fail, e.g.,:
+                    // - vsyscall/kernel page,
+                    // - wild access (NULL or very large bogus address) by app
+                    NOTIFY(1, "virtual2physical translation failure for "
+                           "<%2d, %2d, " PFX">\n",
+                           type, instru->get_entry_size(mem_ref), virt);
+                }
+            }
+        }
+        if (do_write) {
+            // Split up the buffer into multiple writes to ensure
+            // atomic pipe writes.
+            // We can only split before TRACE_TYPE_INSTR, assuming only a few
+            // data entries in between instr entries.
+            if (instru->get_entry_type(mem_ref) == TRACE_TYPE_INSTR) {
+                if ((mem_ref - pipe_start) > ipc_pipe.get_atomic_write_size())
+                    pipe_start = atomic_pipe_write(tid, pipe_start, pipe_end);
+                // Advance pipe_end pointer
+                pipe_end = mem_ref;
+            }
+        }
+    }
+    if (do_write) {
+        // Write the rest to pipe
+        // The last few entries (e.g., instr + refs) may exceed the atomic write size,
+        // so we may need two writes.
+        if ((buf_ptr - pipe_start) > ipc_pipe.get_atomic_write_size())
+            pipe_start = atomic_pipe_write(tid, pipe_start, pipe_end);
+        if ((buf_ptr - pipe_start) > (ssize_t)buf_hdr_slots_size)
+            atomic_pipe_write(tid, pipe_start, buf_ptr);
+    }
+}
+
+static inline ssize_t
+write_trace_data(file_t file, void *data, ssize_t size)
+{
+    ssize_t written = file_ops_func.write_file(file, data, size);
+    DR_ASSERT(written == size);
```

Comment:
This used to be NOTIFY0 with an understandable message and dr_abort, which seems better than "written==size" being printed